### PR TITLE
Tims code updates

### PR DESCRIPTION
**Pull request by dbmurphy**

### CODE REVIEW

File: collector/mongod/oplog_status.go [hunk 1]
```diff
@@ -1,102 +1,116 @@
 package collector_mongod
 
 import (
-    "time"
```
Comment (dbmurphy) [inline collector/mongod/oplog_status.go]: Do we not have a need for time any more?

Comment (timvaillancourt) [inline collector/mongod/oplog_status.go]: No, we get a compile error because it is imported and not used

File: collector/mongod/oplog_status.go [hunk 2]
```diff
@@ -1,102 +1,116 @@
...
+	})
+	oplogStatusSizeBytes = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace:	Namespace,
+		Subsystem:	"replset_oplog",
+		Name:		"size_bytes",
+		Help:		"Size of oplog in bytes",
+	}, []string{"type"})
 )
 
-func GetCollectionSizeMB(db string, collection string, session *mgo.Session) (float64) {
```
Comment (dbmurphy) [inline collector/mongod/oplog_status.go]: Why are we removing this? Displaying it in bits could make it an int64 vs int32 right?

File: collector/mongod/oplog_status.go [hunk 3]
```diff
@@ -1,102 +1,116 @@
...
-    }
-
-    minTime := ParseBsonMongoTsToUnix(results.MinTime)
-    maxTime := ParseBsonMongoTsToUnix(results.MaxTime)
-
-    now := time.Now().Unix()
-    lengthSeconds := maxTime - minTime
-    lengthSecondsNow := now - minTime
-
-    return float64(lengthSeconds), float64(lengthSecondsNow)
```
Comment (dbmurphy) [inline collector/mongod/oplog_status.go]: We are losing some scope here we no longer know now - first time stamp. As discussed before this is causing use to store more data than is needed and more than competing mongo monitoring software stores.

File: collector/mongod/oplog_status.go (hunk 2 above)
Comment (timvaillancourt) [inline collector/mongod/oplog_status.go]: The only lowlevel datatype in prometheus is float64, everything has to be one even if it's a 0/1. It won't accept regular float/int or int64.

File: collector/mongod/replset_status.go [hunk 4]
```diff
@@ -1,232 +1,223 @@
 package collector_mongod
 
 import (
-    "time"
-    "github.com/golang/glog"
-    "github.com/prometheus/client_golang/prometheus"
-    "gopkg.in/mgo.v2"
-    "gopkg.in/mgo.v2/bson"
+	"time"
+
```
Comment (dbmurphy) [inline collector/mongod/replset_status.go]: Why the empty line returns?

File: collector/mongod/replset_status.go [hunk 5]
```diff
@@ -1,232 +1,223 @@
...
-            Name:      "max_n2n_ping_ms",
-            Help:      "Maximum ping in milliseconds to other replica set members",
-    })
+	myState   = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: Namespace,
+		Subsystem: "replset",
+		Name:      "my_state",
+		Help:      "An integer between 0 and 10 that represents the replica state of the current member",
+	}, []string{"set"})
+	term = prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
Comment (dbmurphy) [inline collector/mongod/replset_status.go]: Should we not name this to be more human friendly unless someone know mongo they dont know what term means. Last election_id might be a better name for graphs

File: collector/mongod/replset_status.go [hunk 6]
```diff
@@ -1,232 +1,223 @@
...
+		Subsystem: "replset",
+		Name:      "heatbeat_interval_millis",
+		Help:      "The frequency in milliseconds of the heartbeats",
+	}, []string{"set"})
+	memberHealth = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: Namespace,
+		Subsystem: "replset",
+		Name:      "member_health",
+		Help:      "This field conveys if the member is up (1) or down (0).",
+	}, []string{"set", "name", "state"})
```
Comment (dbmurphy) [inline collector/mongod/replset_status.go]: Again we have scope creep we are add more metrics. we should only monitor what is required, and have meetings on anything  additional

File: collector/mongod/wiredtiger.go [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+		stats.Cache.Export(ch)
+	}
+	if stats.Transaction != nil {
+		stats.Transaction.Export(ch)
+	}
+	if stats.ConcurrentTransactions != nil {
+		stats.ConcurrentTransactions.Export(ch)
+	}
+
+	wtBlockManagerBlocksTotal.Collect(ch)
```
Comment (dbmurphy) [inline collector/mongod/wiredtiger.go]: Does MMS collect all of this? Are we collecting to much? Not enought?

File: collector/mongos/sharding_status.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+			}
+		}
+	}
+	balancerIsEnabled.Set(status.IsBalanced)
+	balancerChunksBalanced.Set(status.BalancerEnabled)
+
+	balancerIsEnabled.Collect(ch)
+	balancerChunksBalanced.Collect(ch)
+	mongosClusterIdHex.Collect(ch)
+	mongosClusterCurrentVersion.Collect(ch)
```
Comment (dbmurphy) [inline collector/mongos/sharding_status.go]: Seems over kill to watch a version with will only change on an upgrade of major release

File: collector/mongos/sharding_status.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	mongosClusterMinCompatVersion.Collect(ch)
+	mongosUpSecs.Collect(ch)
+	mongosPing.Collect(ch)
+	mongosBalancerLockState.Collect(ch)
+	mongosBalancerLockTimestamp.Collect(ch)
```
Comment (dbmurphy) [inline collector/mongos/sharding_status.go]: Not sure this really matters

File: collector/mongod/oplog_status.go (hunk 3 above)
Comment (timvaillancourt) [inline collector/mongod/oplog_status.go]: The previous code had 2 x values for oplog time: "length" and "length to now", which are 2 x float64 values. This new code now returns "head" and "tail" which is also 2 x float64 values that can provide more flexibility in graphing in the future. I feel this is a more optimal approach but this can be rolled back if we must, however this will require updates and retesting of the grafana_mongodb_dashboards project also.

File: collector/mongos/sharding_status.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+		}
+	}
+	balancerIsEnabled.Set(status.IsBalanced)
+	balancerChunksBalanced.Set(status.BalancerEnabled)
+
+	balancerIsEnabled.Collect(ch)
+	balancerChunksBalanced.Collect(ch)
+	mongosClusterIdHex.Collect(ch)
+	mongosClusterCurrentVersion.Collect(ch)
+	mongosClusterMinCompatVersion.Collect(ch)
```
Comment (dbmurphy) [inline collector/mongos/sharding_status.go]: Seems over kill

File: collector/mongos/sharding_status.go [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+				mongosBalancerLockState.With(labels).Set(status.BalancerLock.State)
+			}
+		}
+	}
+	balancerIsEnabled.Set(status.IsBalanced)
+	balancerChunksBalanced.Set(status.BalancerEnabled)
+
+	balancerIsEnabled.Collect(ch)
+	balancerChunksBalanced.Collect(ch)
+	mongosClusterIdHex.Collect(ch)
```
Comment (dbmurphy) [inline collector/mongos/sharding_status.go]: Why do we need the cluster Id graphed it wont change

File: collector/mongod/replset_status.go (hunk 4 above)
Comment (timvaillancourt) [inline collector/mongod/replset_status.go]: replset_status.go is an improved version of my code pulled down from mainline, so this was a choice from the upstream committer, unfortunately. It can be removed

File: collector/mongod/replset_status.go (hunk 5 above)
Comment (timvaillancourt) [inline collector/mongod/replset_status.go]: The area I see highlighted here is "myState" which seems to be a good name to me. This is upstream code so I tried not to mess with it, but I am welcome to change this to anything if it doesn't cause incompatibility with the graph templates

File: collector/mongos/sharding_topology.go [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+)
+
+var (
+	shardingTopoInfoTotalShards = prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace:	Namespace,
+			Subsystem:	"sharding",
+			Name:		"shards_total",
+			Help:		"Total # of Shards in the Cluster",
+	})
+	shardingTopoInfoDrainingShards = prometheus.NewGauge(prometheus.GaugeOpts{
```
Comment (dbmurphy) [inline collector/mongos/sharding_topology.go]: This is over kill also

File: collector/mongod/replset_status.go (hunk 6 above)
Comment (timvaillancourt) [inline collector/mongod/replset_status.go]: replset_status.go is an improved version of my code pulled down from mainline, so this was a choice from the upstream committer. There is no extra work being done to gather these metrics seeing they're already in the result we require, but we can disable anything you think is extraneous if they're not used in graphs and space is a concern

File: collector/mongod/wiredtiger.go (hunk 7 above)
Comment (timvaillancourt) [inline collector/mongod/wiredtiger.go]: I'm not sure on all accounts. I started small and added until I felt like I could "see enough" of what WT was doing. I would say this is the bare minimum we might need, and more can be added.

File: collector/mongos/sharding_status.go (hunk 9 above)
Comment (timvaillancourt) [inline collector/mongos/sharding_status.go]: Removing quite a few of those would be breaking changes to the grafana templates..which ones do you mean, or all?

File: collector/mongos/sharding_status.go (hunk 10 above)
Comment (timvaillancourt) [inline collector/mongos/sharding_status.go]: Removing quite a few of those would be breaking changes to the grafana templates..which ones do you mean, or all?

File: collector/mongos/sharding_status.go (hunk 11 above)
Comment (timvaillancourt) [inline collector/mongos/sharding_status.go]: It's gathered but not graphed. This can be removed if there is concern

File: collector/mongos/sharding_topology.go (hunk 12 above)
Comment (timvaillancourt) [inline collector/mongos/sharding_topology.go]: The # of shards is used in several graphs, and I feel it is useful, no?

File: collector/mongos/sharding_status.go (hunk 8 above)
Comment (timvaillancourt) [inline collector/mongos/sharding_status.go]: Edit: ok, I'll disable the 'Version' section of this code.

File: collector/mongod/replset_status.go (hunk 5 above)
Comment (dbmurphy) [inline collector/mongod/replset_status.go]: I was actually highlighting term = prometheus.NewGaugeVec(prometheus.GaugeOpts{ but differing with due to the beta tag

File: collector/mongos/sharding_topology.go (hunk 12 above)
Comment (dbmurphy) [inline collector/mongos/sharding_topology.go]: No I was talking about  +			Name:		"shards_draining_total", (I had highlighted the  line that was the start of the function, so was not talking about shards_total. Remoing comment we will fix this later as you added the beta tag
